packages/liveblocks-react/src/factory.tsx

```diff
@@ -280,45 +319,44 @@ export function createRoomContext<
     // Note: We'll hold on to the initial value given here, and ignore any
     // changes to this argument in subsequent renders
     const frozen = useInitial({
+      instanceId,
       initialPresence,
       initialStorage,
       unstable_batchedUpdates,
-      shouldInitiallyConnect:
-        shouldInitiallyConnect === undefined
-          ? typeof window !== "undefined"
-          : shouldInitiallyConnect,
+      autoConnect,
     });
 
-    const [room, setRoom] = React.useState<
-      Room<TPresence, TStorage, TUserMeta, TRoomEvent>
-    >(() =>
-      client.enter(roomId, {
+    const [{ room }, setRoomLeavePair] = React.useState(() =>
+      stableEnterRoom(frozen.instanceId, roomId, {
         initialPresence: frozen.initialPresence,
         initialStorage: frozen.initialStorage,
-        shouldInitiallyConnect: frozen.shouldInitiallyConnect,
+        autoConnect: false,
         unstable_batchedUpdates: frozen.unstable_batchedUpdates,
       })
     );
 
     React.useEffect(() => {
-      const room = client.enter<TPresence, TStorage, TUserMeta, TRoomEvent>(
-        roomId,
-        {
-          initialPresence: frozen.initialPresence,
-          initialStorage: frozen.initialStorage,
-          shouldInitiallyConnect: frozen.shouldInitiallyConnect,
-          unstable_batchedUpdates: frozen.unstable_batchedUpdates,
-        }
-      );
+      const pair = stableEnterRoom(frozen.instanceId, roomId, {
```

Comment:
is the idea that this is hitting the cache? I don't totally understand why we have this in the setState as well as here.

Reply:
Yep! In a classic React provider implementation, typically on the first render the context value is still `null`, and then, on mount you would populate the actual context value.

However; we want to populate the `room` with a value on the first render so that we can continue rendering the component tree which may use `useRoom()` which does not expect to see a `null` value, ever.

First, if externally the RoomProvider’s `id` prop get changed, we’ll also want to run the leave function for the old room and connect to the new room. This is what the useEffect and the setState combo are for.

In the most common use case, the room ID prop won’t change, so the following sequence happens:

- Initial render on SSR: `room` has a value (and its state is “initial”)
- Initial hydration render on client: `room` has a value (and its state is “initial”)
- Effect is scheduled for room ID: setter is called, which would normally trigger a re-render, but since the value of `room` is retrieved from cache, its value won’t change and we spare ourselves an extra render. At the same time… we have the opportunity to grab a handle on the `leave` function we didn’t grab on the initial render, and register it for the unmount (or when the room changes)!

This construct already existed before my changes, btw. The previous/current implementation works the same way.